workflow/snakemake_rules/main_workflow.smk

```diff
@@ -265,6 +267,14 @@ rule index_sequences:
             --output {output.sequence_index} 2>&1 | tee {log}
         """
 
+rule get_weights:
+    output: "data/country_week_weights.tsv"
```

Comment:
Re: [using country case counts from OWID](https://github.com/nextstrain/ncov/pull/1106#issuecomment-2101636473)

Trying this with 5ca1afa129268f743e0a76c3337d00364116353e. Trial runs: [open](https://github.com/nextstrain/ncov/actions/runs/9025041404/job/24800039148), [gisaid](https://github.com/nextstrain/ncov/actions/runs/9025045965/job/24800051957)

There's some missing case count data, i.e. sequences are present despite country/week combinations having a case count of 0. Reasons and how these were addressed:

1. Things listed out here:
 
    https://github.com/nextstrain/ncov/blob/5ca1afa129268f743e0a76c3337d00364116353e/scripts/get_weights.py#L33-L44

2. On top of (1), there are also sequences present for weeks and countries missing entirely from case count data. To address this, I added a [patch](https://github.com/nextstrain/augur/commit/60a0f3ed2207c5746aa6fc1aa29ab3f75990cb9f) in the Augur PR that simply drops these sequences rather than erroring out.

Quick tests on the 100k subsampled dataset make me skeptical about this approach. The high case counts attempt to capture more than the available amount of sequences, while at the same time making weights for weeks with small case counts very small and thus not sample from those at all (despite their potential importance in phylogenetic analysis). Hopefully the trial run results will make it clear whether to pursue this direction.

Reply:
Trial runs are complete. Results seem significantly different – it'd be good to have a phylo person take a look. Here's quick links to `gisaid/asia/all-time` for:

1. Current build on `master`: [uniform sampling by division+month](https://next.nextstrain.org/ncov/gisaid/asia/all-time)
2. Original PR implementation: [weighted sampling by country population, uniform sampling by month](https://next.nextstrain.org/staging/ncov/gisaid/trial/victorlin-weighted-sampling-2/asia/all-time)
3. New PR implementation: [weighted sampling by country+week case counts](https://next.nextstrain.org/staging/ncov/gisaid/trial/victorlin-weighted-sampling-3/asia/all-time)

Note from `builds.yaml` that (2) and (3) only apply to the focal Asia samples. Context (non-Asia) samples still conform to (1), but that could be changed.

Reply:
@victorlin First, one big picture comment related to [this comment in the augur PR](https://github.com/nextstrain/augur/pull/1454#issuecomment-2121165641):

> I want to make sure we have at least one solid use case for this feature.

I see example 2 above ("Original PR implementation") as exactly the use case for this feature where we want to minimize [the kind of manual counting and hardcoded subsampling blocks](https://github.com/nextstrain/ncov/commit/9fd01c64f9471638757ab6ce2dd6a43fe090bfeb) that @trvrb had to implement to support proper weighting of India and China relative to the rest of Asia. If we can replace that subsampling YAML approach with a clearer, simpler table of weights by country or region, that's a big win.

Digging into the specific trees you linked to above, there are two main points that came to mind:

1. Weights by country population, uniform sampling by month looks remarkably like the original implementation, but with a simpler config in the ncov repo and slightly better representation of countries like India and China than even the original implementation. Below I've included example views of the map colored by clade per country for each of the three builds linked above where I've also filtered to "region=Asia".

The original view (sampling by division and month):
<img width="675" alt="image" src="https://github.com/nextstrain/ncov/assets/85372/00065a5a-735b-4f86-a7e4-41d455b24366">

The weighted by population view (sampling by weighted country and uniform month):
<img width="647" alt="image 2" src="https://github.com/nextstrain/ncov/assets/85372/879ae867-ef11-4092-a885-0610c2384967">

The weighted by case counts view (sampling by weighted country and case counts):
<img width="675" alt="image 3" src="https://github.com/nextstrain/ncov/assets/85372/1205089a-9782-4ef5-a59e-3b864a18cbaa">

The first two views show how similar the two builds are qualitatively, but the second view has more data for China, India, Indonesia, and Pakistan than the first. The second view has correspondingly fewer samples for Japan, Thailand, and Malaysia. This seems correct. For example, Indonesia and Malaysia have similar numbers of samples in the first view, but Indonesia has 8 times as many people as Malaysia. The second view more accurately represents that difference in population with 194 samples for Indonesia and 23 for Malaysia.

From 31e1099's commit message:

> This requires changing the geographical grouping resolution from division to country, but I assume it was only grouped by division in an attempt to have varying group sizes per country, and that population-based weighting is an acceptable replacement.

I am curious whether @trvrb chose to group by division for this reason only (to get varying group sizes per country) or if he really preferred even sampling by division. Anecdotally (and perhaps I've misremembered the specifics here), I thought he and Marlin were concerned about overrepresentation of sequences from specific divisions in India as new variants were emerging. To explore this case, I made example map views from each of your builds above at the division resolution and filtering to "country=India".

Original view of India by division:
<img width="338" alt="image 4" src="https://github.com/nextstrain/ncov/assets/85372/8a4e6279-afb7-4baf-99cd-04aba45b7c14">

India by division with population weights by country:
<img width="313" alt="image 5" src="https://github.com/nextstrain/ncov/assets/85372/6f3e8364-4aef-43cb-9f8d-695ec1b3e0f8">

India by division with weights by country and case counts:
<img width="313" alt="image 6" src="https://github.com/nextstrain/ncov/assets/85372/ce525a80-4b66-45bd-a598-140c56216213">

The third view confirms what we already saw above, that sampling by case counts enriches for older clades, which I discuss more in my second point below.

The first view shows a similar representation of each state in India, but the second view (which only groups by country) shows an overrepresentation of samples from a smaller set of states. I can't say that either one of these views is "the correct" implementation, since that judgement depends on the question we're trying to answer with these trees. However, this difference in results does raise the question for me of whether would could have weighted subsampling by country _and_ uniform sampling by division in the same group by logic. In other words, could we have `group_by: "country division year month"` and still use the weights per country? If so, we get the best of both worlds for this use case.

2. Weights by case counts do not seem as useful for this specific ncov use case. The tree from example 3 above overrepresents samples from older clades circulating around the start of 2023 (22B) at the expense of including more samples from recent clades (e.g., 22F, 23I). Your first comment in this thread highlights all of the issue with case counts which I think could be summarized as: case counts are noisy data. They are also less reliable data now that there is less funding for surveillance. While population counts per country are also inaccurate and will change over time, the time scale of those changes is decades and you still get useful relative comparisons between countries independent of their genomic surveillance budgets.

This comment above specifically made me worry about using case counts for subsampling in the ncov builds:

> there are also sequences present for weeks and countries missing entirely from case count data. To address this, I added a [patch](https://github.com/nextstrain/augur/commit/60a0f3ed2207c5746aa6fc1aa29ab3f75990cb9f) in the Augur PR that simply drops these sequences rather than erroring out.

When case count data are missing but sequences are present, I would still want to have the sequences in the tree. Even if you reverted this behavior of Augur to not drop sequences when cases were missing, I probably wouldn't reach for case counts as weights in these builds. That said, Miguel might reach for exactly this when designing a subsampling scheme for SARS-CoV-2 in King County where the case counts during a specific time span are more reliable and the analysis is more sensitive to geographic sampling and case load during that time. I would show this example approach to weighting by case counts to him and get his feedback on how useful it would have been for him.

Reply:
Thanks @huddlej, this is really valuable. You confirmed my suspicious - I'll revert 6b6cd2b5cfb96b5a4eaa3a7fdfc9585a6a367e8c.

> could we have `group_by: "country division year month"` and still use the weights per country?

Trying this with 835481da34d7bcda0eda690607730d72eb8763a8 ([GISAID trial run](https://github.com/nextstrain/ncov/actions/runs/9166507128/job/25202025886)). Will update here when the run is complete.

Reply:
Nice! That looks pretty good to me at first glance. The increased number of sequences for India as a whole makes sense from the perspective that each division has to be equally represented by a fixed number of sequences per group within the country and there are a lot of divisions. Looking at the Malaysia/Indonesia representation, I see the same effect where Indonesia has more divisions than Malaysia and ends up with 12 times as many sequences instead of 8 times as many that we'd expect from country population weights alone. This seems like an expected behavior of the group by country and division, though, so I might not worry about this much...

Reply:
Dropped 835481da34d7bcda0eda690607730d72eb8763a8 per https://github.com/nextstrain/ncov/pull/1106#issuecomment-2151177258:

> For ongoing ncov builds, we should be weighting on population size. I wouldn't worry about admin division population size and would just worry about country-level population sizes in terms of weighting.